OAK-9614 : Document best pratices for Oak Access Control Management and Permission Evaluation (#571)
apache/jackrabbit-oak: oak-doc/src/site/markdown/security/authorization/bestpractices.md

-->

Best Practices for Authorization
--------------------------------------------------------------------------------

<!-- MACRO{toc} -->

## Before you get started
### Threat Model

Before you start coding, creating content or setting up access control set aside some time to consider what is needed 
when it comes to securing your application (and what could go wrong). In other words: write a threat model and 
make sure you keep updating it as you continue developing.

The following references provide a good overview as well as guidance on how to build a threat model:

- https://shostack.org/resources/whitepapers/threat-modeling-what-why-how
- https://owasp.org/www-community/Threat_Modeling
- https://owasp.org/www-community/Threat_Modeling_Process

### Content Modelling

As suggested in [Jackrabbbit Wiki](https://jackrabbit.apache.org/archive/wiki/JCR/DavidsModel_115513389.html#DavidsModel-Rule#2:Drivethecontenthierarchy,don'tletithappen)
the content hierarchy in your JCR repository should be designed and access control requirements tend to be a good driver.

Make sure the content design allows for readable and manageable access control setup later on to secure your data. 

> If extra complexity is required, it might indicate problems with your content model. Properly securing your content 

Reviewer: what do you mean with ```extra complexity```?

Author: see above

> secured might subsequently become increasingly hard and prone to mistakes.

Reviewer: I don't understand that sentence :-)

Author: will try to improve :) and provide an example

this case redundant). Start without any access and then keep granting permissions as needed, following the 
[principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege).
In other words: only grant the minimal set of privileges required to perform a particular task.

### Verification

Write tests upfront that verify for each role and task the expected effective permissions (see definition of roles) are 
granted. Neither less nor more.

Ideally, your tests will fail as soon as someone is attempting to make any change to the permission setup.
Granting additional permissions may open up the door for a privilege escalation and revoking permissions will break
your application (if it doesn't you didn't follow the principle of least privilege).

This may also include tests verify that really no permissions are granted at resources that are outside the scope of a 
given role/task

## Oak Specific Best Practices

### Avoid deny

All authorization models present with Apache Jackrabbbit Oak start without any access granted by default i.e. 
implicit deny everywhere. It is therefore recommended to only grant access where needed and avoid adding explicit 

> deny access control entries. In particular in combination with subsequent allows the overall effect will be hard to 

Reviewer: ```suggestion
deny access control entries. In particular in combination with subsequent ```allow```  rules  the overall effect will be hard to 
```

understand as soon as multiple principals are contained in a given subject.

Be wary if you find yourself adding combinations of denies and allows as it might highlight problematic patterns in 

> your content model that will be hard to secure over time.

Reviewer: ```suggestion
your content model that will be hard to understand an secure over time.
```

### Avoid redundancy

Don't specify redundant access control setup just to be on the safe side:

- If access is granted, avoid repeating the same setup down the hierarchy.

> - Avoid setup for principals with administrative access for which permission evaluation is omitted. It might even create a false sense of security.

Reviewer: can you either elaborate what ```principals with administrative access``` are or add a link to such a description?

Author: sure. i will try to clarify and link to the corresponding configuration option

        String toAdd = /content/parent/newchild
        
        Session session = ...
                
        # Testing Privileges
        # ----------------------------------------------------------------------------------------
        
        AccessControlManager acMgr = session.getAccessControlManager();
        Privilege jcrAddChildNodes = acMgr.privilegeFromName(Privilege.JCR_ADD_CHILD_NODES)
        Privilege jcrRemoveChildNodes = acMgr.privilegeFromName(Privilege.JCR_REMOVE_CHILD_NODES)
        Privilege jcrRemoveNode = acMgr.privilegeFromName(JCR_REMOVE_NODE)
        
        # test if (unspecified) child nodes can be added/removed from the parent
        acMgr.hasPrivileges(parentPath, new Privilege[]{jcrAddChildNodes, jcrRemoveChildNodes}
        
        # test if existing child node can be removed
        acMgr.hasPrivileges(toRemove, new Privilege[]{jcrRemoveNode}
        
        
        # Testing Permissions (on the target node NOT on the parent)
        # ----------------------------------------------------------------------------------------

        # test if not-yet existing node could be added at /content/parent/newchild

>         session.hasPermission(toAdd, Session.ACTION_ADD_NODE)

Reviewer: can you add an explicit ```assertTrue(...)``` or ```assertFalse(...)``` to make clear which result you expect? Same for the next statements.

Author: since the example doesn't come with any ac-setup i don't have an expectation on the result itself.... the goal was to illustrate the common confusion between hasPermission and hasPrivileges when it comes to adding/removing nodes.... but i will catch the return value in a hopefully meaningful variable.

Reviewer: to be honest, I have not looked to much into this example. But maybe it might be best if we could expand this to a complete setup to illustrate the difference much better. Or even better, refer to an existing test (not sure if the link above means exactly this).

Author: the exercise linked above will illustrate more or less the same.

   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
-->

Best Practices for Authorization
--------------------------------------------------------------------------------

<!-- MACRO{toc} -->

## Before you get started
### Threat Model

Before you start coding, creating content or setting up access control set aside some time to consider what is needed 
when it comes to securing your application (and what could go wrong). In other words: write a threat model and 
make sure you keep updating it as you continue developing.

The following references provide a good overview as well as guidance on how to build a threat model:

- https://shostack.org/resources/whitepapers/threat-modeling-what-why-how
- https://owasp.org/www-community/Threat_Modeling
- https://owasp.org/www-community/Threat_Modeling_Process

### Content Modelling

As suggested in [Jackrabbbit Wiki](https://jackrabbit.apache.org/archive/wiki/JCR/DavidsModel_115513389.html#DavidsModel-Rule#2:Drivethecontenthierarchy,don'tletithappen)

> the content hierarchy in your JCR repository should be designed and access control requirements tend to be a good driver.

Reviewer: ```suggestion
the content hierarchy in your JCR repository should be designed and not just happen. Access control requirements tend to be a good driver.
```

Make sure the content design allows for a readable and manageable access control setup later on to secure your data. 
Excessive complexity is often a strong indicator for problems with your content model, making its security error prone 
and difficult to reason about (and might ultimately might lead to issues with scaling).

> Here is an example of a access control setup (in Sling RepoInit language) illustrating why content with 

Reviewer: can you add a link to the Sling Repoinit page?

Author: sure! one sec.....

different access requirements should be kept in separate trees and how complexity may yield undesired
effects (see also section 'Remember inheritance' below):

>       # TO BE AVOIDED

Reviewer: I don't understand the scope of this "TO BE AVOIDED". Does it affect the complete example? In that case I would love to see an example where it's done the right way.

Author: yep....

granting your _content-writers_ role full access to just make it work, you probably left your application vulnerable.

- JCR Specification sections [Access Control Management](https://s.apache.org/jcr-2.0-spec/16_Access_Control_Management.html) 
and [Permissions and Capabilities](https://s.apache.org/jcr-2.0-spec/9_Permissions_and_Capabilities.html)
- [Oak Authorization Documentation](../authorization.html) with separate sections for [Access Control Management](../accesscontrol.html) and [Permission Evaluation](../permission.html).
- Exercises for authorization topics below https://github.com/apache/jackrabbit-oak/tree/trunk/oak-exercise/src/test/java/org/apache/jackrabbit/oak/exercise/security/authorization

### Principle of least privilege

Keep in mind that not having any permissions granted is equivalent to denying everything (which is in 
this case redundant). Start without any access and then keep granting permissions as needed, following the 
[principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege).
In other words: only grant the minimal set of privileges required to perform a particular task.

### Verification

> Write tests upfront and verify that for each role and task the expected effective permissions (see definition of roles) are 

Reviewer: Any recommendation how such tests could look like? Examples?

Author: only pseudo-code.... but i can add that

### Avoid redundancy

Don't specify redundant access control setup just to be on the safe side:

- If access is granted, avoid repeating the same setup down the hierarchy.
- Avoid setup for principals with administrative access for which permission evaluation is omitted. It might even create a false sense of security.

### Principal by principle

Oak authorization is designed to work with `java.security.Principal` which is an abstract representation of any kind of 
entity like e.g. individual, a role, a corporation, a login id or even a service.

While JCR specification does not define how the repository knows about principals, Jackrabbit API defines a
[Principal Management](../principal.html) extension.

#### Not every principal is a user/group

Oak allows plugging custom sources of principals which are all reflected through the principal management API.
Therefore, don't assume that every principal is backed by a user or a group. The repository's user management is just 
one potential source of principals.

##### Example : everyone

     # everyone always exists even if there is no such group in the user management
     

>      PrincipalManager principalMgr = ((JackrabbitSession) session).getPrincipalManagere();

Reviewer: ```suggestion
     PrincipalManager principalMgr = ((JackrabbitSession) session).getPrincipalManager();
```

>      Principal everyone = principalManager.getEveryone()

Reviewer: ```suggestion
     Principal everyone = principalManager.getEveryone();
```

#### Membership is no guarantee

Similarly, make sure you always evaluate permissions to verify if a subject has access granted instead of checking if 
a user is member of a group. How access control defined for a particular group principal affects its members is an 
implementation detail of the authorization setup.

##### Example : administrative access

In the default authorization model full access to the repository can be configured for selected user or group principals.
(see [Configuration Parameters](../permission/default.html#configuration) for the default permission evaluation).
If you wish to determine if a given subject has full access, don't assume that there is a group 'administrators' and that 
its members have full access.

#### Stick with group principals

It is preferable to setup access control for group principals instead of individual user principals and then make sure 
your `PrincipalProvider` resolves principal membership according to your needs.

> Further note, that the default authorization model will give precedence to user principals upon evaluation in other words 

Reviewer: ```suggestion
Further note, that the default authorization model will give precedence to user principals upon evaluation. In other words 
```

or user management). The complete list can be found in [Privilege Management : The Default Implementation](../privilege/default.html). 

The minimal set of privileges required for each operation are outlined in [Mapping API Calls to Privileges](../privilege/mappingtoprivileges.html) 
and [Mapping Privileges to Items](../privilege/mappingtoitems.html). 

##### Privileges affecting the parent node

Note in particular for add/removing a node `jcr:addChildNodes` and `jcr:removeChildNodes` are required on 
the parent node respectively i.e. allowing for modification of the child-node collection. 
In addition `jcr:removeNode` needs to be granted on the target node of the removal.

Thus the following subtle difference apply when evaluation effective permissions vs. privileges (see 
also [Permissions vs Privileges](../permission/permissionsandprivileges.html)) and exercises at 
[L4_PrivilegesAndPermissionsTest.java](https://github.com/apache/jackrabbit-oak/blob/trunk/oak-exercise/src/test/java/org/apache/jackrabbit/oak/exercise/security/authorization/permission/L4_PrivilegesAndPermissionsTest.java)):

>         String parentPath = /content/parent

Reviewer: I would prefer if this is formatted as valid java code, including quoting and the semicolon.

Author: vale!

        String toRemove = /content/parent/child
        String toAdd = /content/parent/newchild
        
        Session session = ...
                
        # Testing Privileges
        # ----------------------------------------------------------------------------------------
        
        AccessControlManager acMgr = session.getAccessControlManager();
        Privilege jcrAddChildNodes = acMgr.privilegeFromName(Privilege.JCR_ADD_CHILD_NODES)
        Privilege jcrRemoveChildNodes = acMgr.privilegeFromName(Privilege.JCR_REMOVE_CHILD_NODES)
        Privilege jcrRemoveNode = acMgr.privilegeFromName(JCR_REMOVE_NODE)
        
        # test if (unspecified) child nodes can be added/removed from the parent

>         boolean canModifyChildCollection = acMgr.hasPrivileges(parentPath, new Privilege[]{jcrAddChildNodes, jcrRemoveChildNodes}

Reviewer: ```suggestion
        boolean canModifyChildCollection = acMgr.hasPrivileges(parentPath, new Privilege[]{jcrAddChildNodes, jcrRemoveChildNodes});
```

        
        # test if existing child node can be removed

>         boolean canRemoveNode = acMgr.hasPrivileges(toRemove, new Privilege[]{jcrRemoveNode}

Reviewer: ```suggestion
        boolean canRemoveNode = acMgr.hasPrivileges(toRemove, new Privilege[]{jcrRemoveNode});
```

and [L8_GlobRestrictionTest](https://github.com/apache/jackrabbit-oak/blob/trunk/oak-exercise/src/test/java/org/apache/jackrabbit/oak/exercise/security/authorization/accesscontrol/L8_GlobRestrictionTest.java) 
in the Oak exercise module.

#### Access control setup for system user

If your Oak setup supports principal-based authorization (see [Managing Access by Principal](principalbased.html)) it is 
recommended to leverage it for system sessions associated with OSGi service. It helps to keep application owned 
access control setup apart from regular content.

See also [Service Authentication](https://sling.apache.org/documentation/the-sling-engine/service-authentication.html) in 
Apache Sling.

### Leverage customizations

Oak allows for customization and extensions of all parts of the authorization setup.
If you find yourself struggling to reflect your needs with the built-in functionality, consider extending and customizing
the authorization configuration of the repository.

#### Leverage custom privileges

> If you identify application specific operations that cannot be reflected using the built in privileges, Oak allows to 

Reviewer: ```suggestion
If you identify application specific operations that cannot be reflected using the built-in privileges, Oak allows to 
```

register custom privileges (see section [Privilege Management](../privilege.html#jackrabbit_api)). 

> Note however, that the built-in permission evaluation will not enforce those 

Reviewer: ```suggestion
However, note that the built-in permission evaluation will not enforce those 
```

> custom privileges. Instead you have to enforce it in your application or by writing a custom authorization model 

Reviewer: ```suggestion
custom privileges. Instead you have to enforce it in your application or write a custom authorization model 
```
